### PR TITLE
[MERGE] mail, test_mail: fix composer default values computation

### PR DESCRIPTION
**Pull request by tde-banana-odoo**

### CODE REVIEW

File: addons/test_mail/tests/common.py [hunk 1]
```diff
@@ -12,6 +12,26 @@ class TestMailCommon(MailCommon):
     def _create_channel_listener(cls):
         cls.channel_listen = cls.env['mail.channel'].with_context(cls._test_context).create({'name': 'Listener'})
 
+    @classmethod
+    def _create_records_for_batch(cls, model, count):
+        records = cls.env[model]
+        partners = cls.env['res.partner']
+        country_id = cls.env.ref('base.be').id,
+        for x in range(count):
+            partners += cls.env['res.partner'].with_context(**cls._test_context).create({
```
Comment (awa-odoo) [inline addons/test_mail/tests/common.py:21]: Could build a list of values and then create to take advantage of create_multi.
Or you especially don't want that for some reason?

File: addons/test_mail/tests/common.py [hunk 2]
```diff
@@ -12,6 +12,26 @@ class TestMailCommon(MailCommon):
...
+                'email': '_test_partner_%s@example.com' % (x),
+                'country_id': country_id,
+                'mobile': '047500%02d%02d' % (x, x)
+            })
+            records += cls.env[model].with_context(**cls._test_context).create({
+                'name': 'Test_%s' % (x),
+                'customer_id': partners[x].id,
+            })
+        cls.records = cls._reset_mail_context(records)
+        cls.partners = partners
```
Comment (awa-odoo) [inline addons/test_mail/tests/common.py:32]: Wow...
Can't you access partners from the created records when needed?

Or `return records, partners` and then assign them to "cls" if it's even necessary ?

This is a bit dirty / hidden.

File: addons/test_mail/tests/test_mail_composer.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        }
+        if len(records) == 1:
+            base_context['default_composition_mode'] = 'comment'
+            base_context['default_res_id'] = records.id
+        else:
+            base_context['default_composition_mode'] = 'mass_mail'
+            base_context['active_ids'] = records.ids
+        if add_web:
+            base_context['active_model'] = records._name
+            base_context['active_id'] = records[0].id
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:80]: Should this be inside the `if len(records) == 1:` test?

File: addons/test_mail/tests/test_mail_composer.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        return base_context
+
+
+@tagged('mail_composer')
+class TestComposerForm(TestMailComposer):
+
+    @users('employee')
+    def test_mail_composer_comment(self):
+        composerForm = Form(self.env['mail.compose.message'].with_context(self._get_context(self.test_record, add_web=True)))
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:89]: A wild camelCased variable appears.

File: addons/test_mail/tests/test_mail_composer.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    def test_mail_composer_content_comment_wtpl(self):
+        composer = self.env['mail.compose.message'].with_context({
+            'default_composition_mode': 'comment',
+            'default_model': self.test_record._name,
+            'default_res_id': self.test_record.id,
+        }).create({
+            'template_id': self.template.id,
+        })
+
+        # creation values from from template
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:222]: from from?

File: addons/test_mail/tests/test_mail_composer.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        # creation values are taken
+        self.assertEqual(composer.subject, 'My amazing subject')
+        self.assertEqual(composer.body, '<p>Test Body</p>')
+        self.assertEqual(composer.mail_server_id.id, False)
+
+        # erasing template should erase its content
+        composer.write({'template_id': False})
+        # currently onchange necessary
+        composer.onchange_template_id_wrapper()
+        self.assertEqual(composer.subject, 'My amazing subject')
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:251]: It doesn't erase the subject?

File: addons/test_mail/tests/test_mail_composer.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        composer.write({'template_id': self.template.id})
+        # currently onchange necessary
+        composer.onchange_template_id_wrapper()
+        self.assertEqual(composer.subject, self.template.subject)
+        self.assertEqual(composer.body, self.template.body_html)
+        self.assertEqual(composer.mail_server_id, self.template.mail_server_id)
+
+        # manual values is kept over template
+        composer.write({'subject': 'Back to my amazing subject'})
+        self.assertEqual(composer.subject, 'Back to my amazing subject')
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:266]: This feels a bit duplicate of your previous test unless some specific conditions change?

File: addons/test_mail/tests/test_mail_composer.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+    @users('employee')
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_mail_composer_subject(self):
+        for composition_mode, records in zip(['comment', 'mass_mail'], [self.test_record, self.test_records]):
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:271]: :'(

File: addons/test_mail/tests/test_mail_composer.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+            composer = self.env['mail.compose.message'].with_context(self._get_context(records, add_web=True)).create({})
+            self.assertEqual(composer.composition_mode, composition_mode)
+
+            # default values come from record (comment) or set to False (mass mode)
+            if composition_mode == 'comment':
+                self.assertEqual(composer.record_name, self.test_record.name)
+            else:
+                self.assertFalse(composer.record_name)
+            if composition_mode == 'comment':
+                self.assertEqual(composer.subject, 'Re: %s' % self.test_record.name)
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:281]: Why not group your assertions since the if/else is the same?

**Review by awa-odoo — COMMENTED** [review]
Hello.

Some comments about: "[IMP] test_mail: add new mail composer tests for its internals".

I will check the rest a bit later.

File: addons/mail/wizard/mail_compose_message.py [hunk 10]
```diff
@@ -63,17 +63,25 @@ def default_get(self, fields):
...
+            result['res_id'] = result.get('res_id', self._context.get('active_id'))
+        if 'no_auto_thread' in fields and 'no_auto_thread' not in result and 'model' in fields and (result.get('model') not in self.env or not hasattr(self.env[result['model']], 'message_post')):
             result['no_auto_thread'] = True
 
         vals = {}
         if 'active_domain' in self._context:  # not context.get() because we want to keep global [] domains
             vals['active_domain'] = '%s' % self._context.get('active_domain')
-        if result['composition_mode'] == 'comment':
-            vals.update(self.get_record_data(result))
+        if result.get('composition_mode') == 'comment' and any(fname in fields for fname in ['model', 'res_id', 'partner_ids', 'record_name', 'subject']):
```
Comment (awa-odoo) [inline addons/mail/wizard/mail_compose_message.py:78]: Could use set intersection:
`and set(fields) & set(['model', 'res_id', 'partner_ids', 'record_name', 'subject'])`

But maybe it's a bit less readable, I let you choose.

File: addons/mail/wizard/mail_compose_message.py [hunk 11]
```diff
@@ -63,17 +63,25 @@ def default_get(self, fields):
...
+            record_values = self.get_record_data(result)
+            vals.update(dict(
```
Comment (awa-odoo) [inline addons/mail/wizard/mail_compose_message.py:80]: I slightly prefer:

```py
vals.update({
    fname: value
    for fname, value in record_values.items()
    if fname in fields
})
```

But it's a matter of taste at this point.

File: addons/test_mail/tests/common.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/common.py:32]: Well that's a move from other tests. I agree this is dirty and hidden but as it targets stable, I will keep it as it is, even for a test module.

**Review by awa-odoo — COMMENTED** [review]
Looks fine, I made some code suggestions if you feel motivated :)

File: addons/test_mail/tests/common.py (hunk 1 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/common.py:21]: It was copy pasted. I suppose a create multi could work indeed. Will try.

Comment (tde-banana-odoo) [reply addons/test_mail/tests/common.py:21]: Et bonjour hein.

File: addons/test_mail/tests/test_mail_composer.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py:80]: When testing various actions I often have both active_id and active_ids in context ... not sure. Anway I don't think it has a big impact as if composer behaves well with both of them, one of them is ok.

File: addons/test_mail/tests/test_mail_composer.py (hunk 4 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py:89]: :ghost:

File: addons/test_mail/tests/test_mail_composer.py (hunk 5 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py:222]: Indeed indeed.

File: addons/test_mail/tests/test_mail_composer.py (hunk 6 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py]: Some strange behavior indeed... I am asserting current behavior anyway.

File: addons/test_mail/tests/test_mail_composer.py (hunk 7 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py]: Merged tests.

File: addons/test_mail/tests/test_mail_composer.py (hunk 8 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py]: :)

File: addons/test_mail/tests/test_mail_composer.py (hunk 9 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer.py]: Cleaned.

File: addons/mail/wizard/mail_compose_message.py [hunk 12]
```diff
@@ -63,17 +63,25 @@ def default_get(self, fields):
             if missing_author:
                 result['author_id'] = author_id
 
-        result['composition_mode'] = result.get('composition_mode', 'comment')
-        result['model'] = result.get('model', self._context.get('active_model'))
-        result['res_id'] = result.get('res_id', self._context.get('active_id'))
-        if 'no_auto_thread' not in result and (result['model'] not in self.env or not hasattr(self.env[result['model']], 'message_post')):
+        if 'composition_mode' in fields and 'composition_mode' not in result:
+            result['composition_mode'] = result.get('composition_mode', 'comment')
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:67]: Actually I removed the line completely, as there is a default on the field itself.

File: addons/mail/wizard/mail_compose_message.py [hunk 13]
```diff
@@ -168,7 +165,6 @@ def get_record_data(self, values):
         if subject and not (subject.startswith('Re:') or subject.startswith(re_prefix)):
             subject = "%s %s" % (re_prefix, subject)
         result['subject'] = subject
-
```
Comment (awa-odoo) [inline addons/mail/wizard/mail_compose_message.py:171]: A bit of unnecessary diff left here.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff
